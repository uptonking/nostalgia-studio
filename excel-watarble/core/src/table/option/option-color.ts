import { ValueObject } from '@datalking/pivot-entity';

import {
  type ICreateOptionColorSchema,
  type IOptionColor,
  type IOptionColorName,
  type IOptionColorShade,
} from './option.schema';
import { optionColorOrder } from './option.schema';

export class OptionColor extends ValueObject<IOptionColor> {
  get name() {
    return this.props.name;
  }

  get shade() {
    return this.props.shade;
  }

  static get defaultShade(): IOptionColorShade {
    return 5;
  }

  static get defaultColorName(): IOptionColorName {
    return 'indigo';
  }

  static create(input?: ICreateOptionColorSchema): OptionColor {
    return input
      ? new this({
          name: input.name ?? this.defaultColorName,
          shade: input.shade ?? this.defaultShade,
        })
      : this.defaultColor;
  }

  static get defaultColor(): OptionColor {
    return new this({
      name: this.defaultColorName,
      shade: this.defaultShade,
    });
  }

  static createColors(
    inputs: (ICreateOptionColorSchema | undefined)[],
  ): OptionColor[] {
    const colors: OptionColor[] = [];

    for (const [index, input] of inputs.entries()) {
      if (!input && colors[index - 1]) {
        colors.push(colors[index - 1].next());
      } else {
        colors.push(OptionColor.create(input));
      }
    }

    return colors;
  }

  static nextColorName(colorName?: IOptionColorName): IOptionColorName {
    if (!colorName) return optionColorOrder[0];

    const index = optionColorOrder.indexOf(colorName);
    const nextColorIndex =
      index === optionColorOrder.length - 1 ? 0 : index + 1;
    const nextColorName = optionColorOrder[nextColorIndex];

    return nextColorName;
  }

  next(): OptionColor {
    const nextColorName = OptionColor.nextColorName(this.name);
    const shade = this.shade;

    return new OptionColor({
      name: nextColorName,
      shade,
    });
  }
}
