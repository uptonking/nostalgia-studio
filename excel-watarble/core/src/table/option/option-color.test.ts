import { OptionColor } from './option-color';

describe('create', () => {
  test('should create default option color', () => {
    const oc = OptionColor.create();
    expect(oc).to.be.instanceOf(OptionColor);
    expect(oc.name).to.be.eq('indigo');
    expect(oc.shade).to.be.eq(5);
  });

  test('should create option color with value', () => {
    const oc = OptionColor.create({ name: 'dark', shade: 0 });
    expect(oc).to.be.instanceOf(OptionColor);
    expect(oc.name).to.be.eq('dark');
    expect(oc.shade).to.be.eq(0);
  });

  test('should create option color with only name', () => {
    const oc = OptionColor.create({ name: 'dark' });
    expect(oc).to.be.instanceOf(OptionColor);
    expect(oc.name).to.be.eq('dark');
    expect(oc.shade).to.be.eq(5);
  });

  test('should create option color with only shade', () => {
    const oc = OptionColor.create({ shade: 0 });
    expect(oc).to.be.instanceOf(OptionColor);
    expect(oc.name).to.be.eq('indigo');
    expect(oc.shade).to.be.eq(0);
  });
});

describe('next', () => {
  test('should get next color', () => {
    const color = OptionColor.create();
    expect(color.next()).toMatchInlineSnapshot(`
      OptionColor {
        "props": {
          "name": "blue",
          "shade": 5,
        },
      }
    `);
  });

  test('should get next color for last color name', () => {
    const color = OptionColor.create({ name: 'orange' });
    expect(color.next().name).to.be.eq('dark');
  });
});

describe('createColors', () => {
  test('should create colors if no input', () => {
    const colors = OptionColor.createColors([undefined, undefined]);
    expect(colors).to.toMatchSnapshot();
  });

  test('should create colors if has input', () => {
    const colors = OptionColor.createColors([
      { name: 'blue' },
      { name: 'red' },
    ]);
    expect(colors).to.toMatchSnapshot();
  });
});
