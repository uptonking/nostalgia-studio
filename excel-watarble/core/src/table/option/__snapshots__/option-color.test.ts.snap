// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createColors > should create colors if has input 1`] = `
[
  OptionColor {
    "props": {
      "name": "blue",
      "shade": 5,
    },
  },
  OptionColor {
    "props": {
      "name": "red",
      "shade": 5,
    },
  },
]
`;

exports[`createColors > should create colors if no input 1`] = `
[
  OptionColor {
    "props": {
      "name": "indigo",
      "shade": 5,
    },
  },
  OptionColor {
    "props": {
      "name": "blue",
      "shade": 5,
    },
  },
]
`;
